[SPARK-49243][INFRA][K8S] Add `OpenContainers` Annotations in docker images

dongjoon-hyun opened the pull request:
### What changes were proposed in this pull request?

This PR aims to add `OpenContainers` Annotations to docker image to add metadata like `Apache License`.

- https://specs.opencontainers.org/image-spec/annotations/

### Why are the changes needed?

**BEFORE**
```
$ docker inspect apache/spark:3.5.2 | jq '.[0].Config.Labels'
{
  "org.opencontainers.image.ref.name": "ubuntu",
  "org.opencontainers.image.version": "20.04"
}
```

**AFTER**

```
$ NO_MANUAL=1 ./dev/make-distribution.sh -Pkubernetes
$ bin/docker-image-tool.sh -t SPARK-49243 -n build
$ docker inspect spark:SPARK-49243 | jq '.[0].Config.Labels'
{
  "org.opencontainers.image.authors": "Apache Spark project <dev@spark.apache.org>",
  "org.opencontainers.image.licenses": "Apache-2.0",
  "org.opencontainers.image.ref.name": "Apache Spark Scala/Java Image",
  "org.opencontainers.image.version": ""
}
```

### Does this PR introduce _any_ user-facing change?

No, it's only a metadata change.

### How was this patch tested?

Manual review.

### Was this patch authored or co-authored using generative AI tooling?

No.

dongjoon-hyun committed:
[SPARK-49243][INFRA][K8S] Add `OpenContainers` Annotations in docker images

dongjoon-hyun commented:
Could you review this PR, @viirya ?

viirya reviewed (approved):
Looks good to me. Thanks @dongjoon-hyun

dongjoon-hyun commented:
Thank you, @viirya !

yaooqinn commented on dev/create-release/spark-rm/Dockerfile:
```diff
@@ -17,6 +17,10 @@
 
 # Image for building Spark releases. Based on Ubuntu 22.04.
 FROM ubuntu:jammy-20240227
+LABEL org.opencontainers.image.authors="Apache Spark project <dev@spark.apache.org>"
```
Apache Spark Community?

dongjoon-hyun replied:
> Apache Spark Community?

Isn't it too broad because it include a read-only users.

yaooqinn replied:
I believe it's acceptable to include read-only users because non-code contributions, such as testing and sharing use cases, should also be acknowledged.
The [User](https://apache.org/foundation/how-it-works/#users) also plays a role in contributing according to 'The Apache Way'.

dongjoon-hyun replied:
> I believe it's acceptable to include read-only users because non-code contributions, such as testing and sharing use cases, should also be acknowledged. The [User](https://apache.org/foundation/how-it-works/#users) also plays a role in contributing according to 'The Apache Way'.

I got your point. It seems that there was a little misunderstanding between you and me.
- When I say `read-only users`, I embraced `read-only users` (who didn't contribute at all in all ways) as `Apache Spark Community`.

The point is that I have a broader concept on the community than yours.

The next question is that 
- `Contributors` are `Authors`?

dongjoon-hyun replied:
One more thing. I believe I used the term `Project` in the author in this way.
- https://apache.org/projects/#what-is-an-open-source-project

> An open source project includes all aspects of creating, maintaining, and distributing open source software including community building and mentoring, communication, the release process, and everything in between.

yaooqinn replied:
Shall we capitalize 'project' to make it consistent with others

dongjoon-hyun replied:
Maybe, did you mean `Apache Spark` by `with others`? It's possible.

However, technically, `Apache Spark™` is a trademark and the unique name of our project while `project` is a normal noun.

[
![Screenshot 2024-08-15 at 08 11 04](https://github.com/user-attachments/assets/8598f62c-64b1-42c2-997d-720a613cc48d)
](url)

Here is a sample example sentence according to ASF guideline. Unlike `Apache Spark`, the following normal noun like `software` is used in a lower case. You can see that the phrase, `the Apache Foo project`, in the same document as a example.
https://www.apache.org/foundation/marks/#guidelines

> "Free copies of Apache ProjectName software under the Apache License and support services for Apache ProjectName are available at my own company website."

yaooqinn replied:
Oh, I mean 'Apache Spark ~p~Project' like 'S'cala, 'I'mage etc, but it's just a nit

yaooqinn commented on dev/create-release/spark-rm/Dockerfile:
```diff
@@ -17,6 +17,10 @@
+LABEL org.opencontainers.image.authors="Apache Spark project <dev@spark.apache.org>"
+LABEL org.opencontainers.image.licenses="Apache-2.0"
+LABEL org.opencontainers.image.ref.name="Apache Spark Release Manager Image"
+LABEL org.opencontainers.image.version=""
```
Can it be removed if it's blank?

dongjoon-hyun replied:
Ya, I tried, but we should have a blank here to remove the previous OS layer version as described in the PR description.

```
$ docker inspect apache/spark:3.5.2 | jq '.[0].Config.Labels'
{
  "org.opencontainers.image.ref.name": "ubuntu",
  "org.opencontainers.image.version": "20.04"
}
```

dongjoon-hyun replied:
Without this blank, the information could be interpreted as `Apache Spark 20.04`.

yaooqinn replied:
Got it. Thanks for the explanation.
Maybe we can add some comment here.

dongjoon-hyun replied:
Sure. I will add comments, @yaooqinn .

Ya, I agree that the AS-IS is obscure. 

So, I'm currently trying to improve this area including `bin/docker-image-tool.sh` via version detection from `RELEASE`. For the K8s images, I guess I can deliver a better way to fill this field. But, for this `spark-rm/Dockerfile`, I guess it's not worth to do.

yaooqinn replied:
> via version detection from RELEASE

+1

yaooqinn commented on resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile:
```diff
@@ -18,6 +18,10 @@
 FROM ${base_img:-spark}
+LABEL org.opencontainers.image.authors="Apache Spark project <dev@spark.apache.org>"
+LABEL org.opencontainers.image.licenses="Apache-2.0"
+LABEL org.opencontainers.image.ref.name="Apache Spark Python Image"
```
PySpark?

dongjoon-hyun replied:
Ya, I also considered `Apache Spark PySpark Image` and `Apache Spark SparkR Image`.

But, I believe `Python` would be better because we have `Apache Spark Scala/Java Image` as a base image.

In addition, I'm not sure, but `Apache Spark Go Image` comes later. So, for consistency, language name could work extensively.

dongjoon-hyun replied:
Also, please see our docker hub. The Docker convention in the name is languages like `scala`, `r`, `python3` instead of `SparkR` or `PySpark`.

![Screenshot 2024-08-14 at 23 33 02](https://github.com/user-attachments/assets/a879d620-7f68-4a08-a201-830f881783ea)

yaooqinn replied:
I was considering whether it will conflict with the https://pypi.org/project/pyspark-connect/ or not.

But it looks fine to me now

dongjoon-hyun committed:
Address comments by adding a comment

dongjoon-hyun commented:
Thank you for thorough review, @yaooqinn .

I addressed your comments by adding a comment line and replied for the discussion part here.
- https://github.com/apache/spark/pull/47766#discussion_r1718042773
- https://github.com/apache/spark/pull/47766#discussion_r1718054277

Please let me know your opinion.

yaooqinn reviewed (approved):
LGTM

dongjoon-hyun commented:
Thank you! Merged to master.
